rfc/rfc-50/rfc-50.md

```diff
@@ -0,0 +1,93 @@
+
+### Problem description
+
+- HUDI designs the Timeline service for processing and caching when accessing metadata , but currently does not converge all access to metadata to the Timeline service, such as the acquisition of a complete timeline.
+- When the number of tasks written increases, a large number of repeated access to metadata will lead to high HDFS NameNode requests, causing greater pressure and not easy to expand.
+
+### Spark and Flink write flow comparison diagram
+
+Since Hudi is designed based on the Spark micro-batch model, in the Spark write process, all operations on the timeline are completed on the driver side, and then distributed to the executor side to start the write operation.
+
+But for Flink , Write tasks are resident services due to their pure streaming model. There is also no highly reliable communication mechanism between the user-side JM and the TM in Flink, so the TM needs to obtain the latest instant by polling the timeline for writing.
+
+![](ComparisonDiagram.png)
+
+### Current
+
+![](CurrentDesign.png)
+
+The current design implementation has two main problems with the convergence timeline
+- Since the timeline of the task is pulled from the Embedded timeline service, the refresh mechanism of the Embedded timeline service itself will doesn't work
+- MetaClient and HoodieTable are decoupled. Obtain the timeline in MetaClient and then request the Embedded timeline service to obtain file-related information through the FileSystemViewManager in HoodieTable combined with the timeline. There are circular dependencies and problems in the case of using MetaClient alone without creating HoodieTable
+
+## Implementation
+
+### Design target
+
+The goal of this solution is to converge the acquisition of timelines and obtain them through the Embedded timeline service uniformly. The timeline is pulled through HDFS only when the Embedded timeline service is not started.
+
+### Converge the request to loop instant in Flink to JM
+
+- Store the latest instant on the Embedded Timeline Server. Every time JM modifies the instant state, it actively performs a sync to Embedded Timeline Server
+- Return the latest instant directly when the task pulls the latest instant
+
+### Converge the request to pull instant in meta  client initialization to JM
+
+- Abstract the timeline-related acquisition methods into the new interface TableTimelineView, and create the corresponding TimelineViewManager in MetaClient, and obtain the timeline through TimelineViewManager.
+
+![](Design.png)
+
+### Flink optimization before and after schematic diagram
+
+![](SchematicDiagram.png)
+
+## Rollout/Adoption Plan
+
+- What impact (if any) will there be on existing users?
+    - Since the Embedded Timeline Service is used to pull the timeline, users who use flink to write to hudi will observe that file system requests are greatly reduced, thereby reducing the pressure on the file system. 
```

Comment:
can you also put a note here to analyze the impact to spark users? since you also mentioned below planning to remove the existing behavior.

Reply:
Okay, This will not affect the Spark user, because spark only get the timeline from the driver, and the behavior of the timeline accessed by the driver is almost the same as before.